### PR TITLE
added num image recomputation on slotchange

### PR DESCRIPTION
**Pull request by billmoser**
this allows user-defined polymer elements to include an image-carousel, and populate it using dom-repeat.  An element to demonstrate this, image-carousel-demo-app.html, had been added, and demo/index.html updated to show its usage.  

Once this has been accepted & merged, I'll create an additional pull request for a selection notification.

### CODE REVIEW

File: image-carousel.html [hunk 1]
```diff
@@ -275,6 +275,10 @@
 
       connectedCallback() {
         super.connectedCallback();
+        // recompute number of images when slot contents changes
+        this.shadowRoot.querySelector('slot[attribute="images"]').addEventListener('slotchange', e => {
+            this.computeNumberOfImages();
```
Comment (a1626) [inline image-carousel.html]: Indentation of project is `two spaces`. Can you change it to 2 spaces here? Or you can also write it in one single line like

`        this.shadowRoot.querySelector('slot[attribute="images"]').addEventListener('slotchange', e => this.computeNumberOfImages());
`

File: image-carousel.html [hunk 2]
```diff
@@ -275,6 +275,10 @@
 
       connectedCallback() {
         super.connectedCallback();
+        // recompute number of images when slot contents changes
```
Comment (a1626) [inline image-carousel.html:282]: Sorry to ask but will it be `slot content changes` or `slot contents changes`?

File: demo/image-carousel-demo-app.html [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+          // used to pass in images for the dom-repeat
+          images: {
+            type: Array
+            }
+          }
+      }
+    }
+    window.customElements.define(ImageCarouselDemoApp.is, ImageCarouselDemoApp);
+  </script>
+</dom-module>
```
Comment (a1626) [inline demo/image-carousel-demo-app.html:29]: Can you fix indentation of this file also

File: demo/index.html [hunk 4]
```diff
@@ -128,5 +132,30 @@ <h3 id="no-icons">image-carousel demo with no navigation icons</h3>
...
+              "https://c2.staticflickr.com/2/1217/4607963354_e1a8fea210_z.jpg",
+              "https://cs1.livemaster.ru/storage/e6/45/00b119c13ed9f8b4dc6363a60d--materialy-dlya-tvorchestva-nabor-dlya-vyshivki-biserom-belyj-ti.jpg",
+              "https://tse4.mm.bing.net/th?id=ORT.TH_470633631&pid=1.12&eid=G.470633631"]'>
+            </image-carousel-demo-app>            
+          </template>
+        </dom-bind>
+        <div></div>
+      </demo-snippet>
+    </div>
+
```
Comment (a1626) [inline demo/index.html]: Can you follow the pattern of the rest of the examples i.e. 3 new lines before the example and no new line between closing tag of `div` and `body`.

**Comment by a1626** [comment]
Can you add new test cases also?

File: demo/image-carousel-demo-app.html [hunk 5]
```diff
@@ -0,0 +1,31 @@
+<link rel="import" href="../bower_components/polymer/polymer-element.html">
+<link rel="import" href="../image-carousel.html">
+<link rel="import" href="../bower_components/polymer/lib/elements/dom-repeat.html">
```
Comment (a1626) [inline demo/image-carousel-demo-app.html]: Replace `bower_component` with `..` as `polymer serve` should be able to find the correct path.

**Review by a1626 — CHANGES_REQUESTED** [review]

**Comment by a1626** [comment]
Also can you remove the listener in `disconnectedCallback `

File: image-carousel.html (hunk 2 above)
Comment (billmoser) [reply image-carousel.html:282]: I think either way works, depending on how you think about it, but whichever you wish is fine with me.

File: demo/image-carousel-demo-app.html (hunk 3 above)
Comment (billmoser) [reply demo/image-carousel-demo-app.html:29]: done.

File: demo/image-carousel-demo-app.html (hunk 5 above)
Comment (billmoser) [reply demo/image-carousel-demo-app.html]: done.

File: demo/index.html (hunk 4 above)
Comment (billmoser) [reply demo/index.html]: done.

File: image-carousel.html (hunk 1 above)
Comment (billmoser) [reply image-carousel.html]: done.

**Comment by billmoser** [comment]
I added a test case, but its very simple.  I think I should create a fixture which instantiates image-carousel-demo-app and populates the slot in the fixture, rather than using the addLightDom() method.  That way, I can then query for the number of images to see if the slot was filled correctly.  I assume that it's possible to do so -- what do you think?

**Comment by a1626** [comment]
Ya, i think your test case should also make sure that the event was fired and right information was passed down.

**Comment by billmoser** [comment]
I've updated the test to check that the number of images is correct (should be 3).  Also verified that this was actually a bug by commenting out the fix, which caused the test to fail with _numberOfImages=0.

File: demo/image-carousel-demo-app.html [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<link rel="import" href="../polymer/polymer-element.html">
```
Comment (a1626) [inline demo/image-carousel-demo-app.html]: This and line 3 will be `../../` instead of `../` for `polymer-cli` to correctly map it.

File: image-carousel.html [hunk 7]
```diff
@@ -172,6 +172,10 @@
      */
 
      class ImageCarousel extends Polymer.GestureEventListeners(Polymer.Element) {
+      constructor() {
+        super();
+        this._slotChangeHandler = this.computeNumberOfImages.bind(this);
```
Comment (a1626) [inline image-carousel.html:177]: Why is this required? why not use `computeNumberOfImages` directly in `eventListener`?

**Review by a1626 — CHANGES_REQUESTED** [review]

**Review by a1626 — APPROVED** [review]

**Comment by a1626** [comment]
Left some minor comments. Rest of it looks good.

File: image-carousel.html (hunk 7 above)
Comment (billmoser) [reply image-carousel.html:177]: The problem is in that the disconnectedCallback() requires a reference to the function as well, otherwise you'll have a memory leak.  Using
```
removeEventListener('slotchange', this.computeNumberOfImages.bind(this));
```
will not work, since
```
this.computeNumberOfImages.bind(this) !== this.computeNumberOfImages.bind(this)
```
trust me --- I tried that in a previous commit :)

File: demo/image-carousel-demo-app.html (hunk 6 above)
Comment (billmoser) [reply demo/image-carousel-demo-app.html]: Ok, just updated this.  For my own edification, what goes wrong with the way it was? I'm just using 'polymer build' and 'polymer serve', and it _seemed_ to be working fine.  What am I missing?
